### ISSUE TITLE
Streaming data from Azure to local file taking a lot of time compared to using Jclouds

### ISSUE THREAD

**Issue by noviceamateur** [gaul/s3proxy#275]
I am using S3Proxy to download a file of size 150 MB to my local storage. 
I used s3 java sdk to make calls and changed the endpoint to the s3proxy endpoint which is running locally.

This is the code I am using to download the file ->

        GetObjectRequest getObjectRequest = new GetObjectRequest(bucketname , filename);
        final Instant t1 = Instant.now();
        S3Object s3Object = s3.getObject(getObjectRequest);
        final Instant t2 = Instant.now();
        System.out.println("Object Get time  time = " + Duration.between(t1, t2).toMillis());
        S3ObjectInputStream s3ObjectInputStream = s3Object.getObjectContent();
        final Instant t3 = Instant.now();
        writeToFile(s3ObjectInputStream, filename);
        final Instant t4 = Instant.now();
        System.out.println("Download file time = " + Duration.between(t3, t4).toMillis());

Here is the function writeToFile which writes the stream to a file. 

    static void writeToFile(InputStream inputStream, String filename) throws IOException {

        File file = new File(filePath + filename);
        if (!file.createNewFile()) {
            file.delete();
            file.createNewFile();
        }
        java.nio.file.Files.copy(inputStream, file.toPath(), StandardCopyOption.REPLACE_EXISTING);

    }

This is the output -> 
Object Get time = 171300
Download file time = 534

Why does the get object call takes so long ? That call should only return the stream from where we can download the file right?

I did the same experiment by using Jclouds API directly to download the same file and download it locally.
Here is the code I used -> 

        BlobStoreContext context = ContextBuilder.newBuilder("azureblob")
                                                 .credentials(System.getenv("AZURE_ACCESS_ID"), System.getenv
                                                         ("AZURE_ACCESS_KEY"))
                                                 .buildView(BlobStoreContext.class);
        BlobStore blobStore = context.getBlobStore();
        final Instant t1 = Instant.now();
        Blob blob = blobStore.getBlob(containerName, filename);
        final Instant t2 = Instant.now();
        System.out.println("Object Get time  time = " + Duration.between(t1, t2).toMillis());
        InputStream inputStream = blob.getPayload().openStream();
        final Instant t3 = Instant.now();
        writeToFile(inputStream, filename);
        final Instant t4 = Instant.now();
        System.out.println("Download file time " + Duration.between(t3, t4).toMillis());

Here is the output for this code sample ->
Object Get time = 3750
Download file time = 270776

This looks like an acceptable time and comparable to the time which azure sdk provides.

My question is why s3 get object call taking so much longer than the actual streaming time when using s3proxy?

**Comment by noviceamateur** [comment]
Yes. It is invalid. We made changes in Jclouds to use Async http calls using Apache Async Http client and for some reason it is reading the entire file in local memory before calling the completed callback.

**Comment by noviceamateur** [comment]
But Jclouds doesnt provide async API support right ?
These are the changes we made to our local fork so that it uses Apache Async http client.

Jclouds sync calls are using HttpURLConnection client right ?
So, I tested using the original code. This client doesnt buffer the response locally before calling the callback.

The problem is with the implementation of Apache Async Client here. It uses a basic response consumer which buffers the response before calling the completed callback.
